pytorch_lightning/trainer/supporters.py

```diff
@@ -120,26 +120,134 @@ def mean(self):
 
 
 class PredictionCollection(object):
-    def __init__(self, global_rank: int, world_size: int):
+
+    ID_KEY = 'id'
+
+    def __init__(self, global_rank: int, world_size: int, all_gather_fn: Callable):
         self.global_rank = global_rank
         self.world_size = world_size
-        self.predictions = {}
-        self.num_predictions = 0
+        self.all_gather_fn = all_gather_fn
+        self._legacy_predictions = {}
+        self._predictions = {stage: {} for stage in LoggerStages}
+
+    @property
+    def predictions(self) -> Dict:
+        return self._predictions[self.current_stage]
+
+    def _cache_prediction(self, predictions: List[Dict], dl_idx: int) -> None:
+        cache = self.predictions
+
+        cache.setdefault(dl_idx, {})
+
+        for pred in predictions:
+            if self.ID_KEY in pred:
+                key = pred[self.ID_KEY]
+                if not isinstance(key, (int, float)):
+                    raise MisconfigurationException(
+                        f"`id` key should be either a int or float. Found {type(key)}."
+                    )
+            else:
+                raise MisconfigurationException(
+                    f"The predictions dict requires an `{self.ID_KEY}` key."
+                )
+
+            if key in cache[dl_idx]:
+                raise MisconfigurationException(
+                    "Prediction Collection doesn't support multiple predictions for one sample yet.")
+
+            cache[dl_idx][key] = pred
+
+    def cache(self, predictions: List[Dict], dl_idx: int, current_stage: str) -> None:
+        """
+        This function expects predictions to be a list of dictionnaries.
+        Each dictionary should contain an unique key `id`.
+        The `id` key-value should be a unique number to identify each sample.
+
+        Example::
+
+            indexes = ...
+            preds = ...
+
+            # we need an unique number id to identify each sample.
+            predictions = [
+                {"id": id, "prediction": prediction}
+                for id, prediction in zip(indexes, preds)
+            ]
+
+            self.add_predictions(predictions)
+        """
+        if predictions is None or (isinstance(predictions, list) and len(predictions) == 0):
```

Comment:
```suggestion
        if predictions is None or (isinstance(predictions, list) and not predictions):
```
eventually, it can be pure `if not predictions` which covers None and all empty iterators